### DISCUSSION TITLE
Slashdash comment spec

### DISCUSSION THREAD

**Discussion by lilyball** [kdl-org/kdl#29] [Q&A] [answered]
The spec right now includes

```
ws := bom | ' ' | '\t' | multi-line-comment | slashdash-comment
slashdash-comment := '/-' (node | value | prop | node-children)
```

The problem is this makes the following perfectly legal:

```
foo /-{
    embedded-node "weirdness"
} "value"
```

Which is to say, I've now inserted a token somewhere it doesn't belong. What's more, I can also write something like

```
foo "bar" /-one /-two /-three {
  three-children
}
{
  two-children
}
{
  one-children
}
{
  foo-children
}
```

The issue here is that if I have a slashdash, I can now insert any of the valid commentable token types anywhere whitespace is allowed. And because this includes a full node, and nodes will terminate on a braced block, I get this weird chain of braced blocks at the end. I can even do stuff with comments, like

```
foo "bar" /-one // comment
continuation="of foo"
```

or stuff with esclines

```
foo "bar" /-one \
"two" \ // we're still in one
"this is also in one" // comment to terminate one
"we're finally back to foo"
```

And there's a potential parsing ambiguity here, as the definition of `slashdash-comment` puts `node` and `value` at the same spot, so if I write `foo /-true { child }` it's not clear whether the `true` will be treated as a node title (and therefore the child will be part of the comment) or whether it will be treated as a boolean value (and therefore the child will belong to `foo`).

I think fixing this will require splitting `slashdash-comment` up into a separate rule per token type, and then allowing this in each token. So we'd end up with something like

```
node := '/-'? identifier (node-space node-argument)* (node-space node-children)? single-line-comment?
node-argument := '/-'? ( prop | value )
node-children := '/-'? '{' nodes '}'
```

This way only tokens that are valid for the current position can be slashdashed.

**Comment by zkat** [comment]
Oh hey. Would you like to PR this change so it gets properly credited? 👍🏻

**Reply by lilyball** [reply]
Sure, I can do that. I'll have to revisit this later this evening though.

**Reply by zkat** [reply]
take your time! Thanks for taking a look and finding these weird corner-cases. This is fantastic!

**Comment by lilyball** [comment, marked as answer]
Fixed by #33.
